### PR TITLE
sstables/mx/reader: coroutinize some methods of mx_sstable_mutation_reader

### PR DESCRIPTION
**Pull request by tchaikov**
just for better readability.

### CODE REVIEW

File: sstables/sstable_set.cc [hunk 1]
```diff
@@ -1366,14 +1366,12 @@ sstable_set::make_local_shard_sstable_reader(
...
-            return sst->make_reader(s, permit, pr, slice, trace_state, fwd, fwd_mr, monitor_generator(sst));
-        };
+        auto reader = sst->make_reader(s, permit, pr, slice, trace_state, fwd, fwd_mr, monitor_generator(sst));
         // Auto-closed sstable reader is only enabled in the context of fast-forward to partition ranges
         if (!fwd && fwd_mr) {
-            return make_auto_closed_sstable_reader(sst, make_reader(), permit);
+            return make_auto_closed_sstable_reader(sst, std::move(reader), permit);
         }
-        return make_reader();
+        return reader;
```
Comment (denesb) [inline sstables/sstable_set.cc:1374]: Wasn't this extracted to a separate PR?

**Review by denesb — COMMENTED** [review]
I'm not sure we should do this: it introduces an unconditional allocation each time these methods are called and they are on a performance-sensitive path. Not super sensitive, as we are reading from the disk anyway.

I want a second opinion from the guardian of allocations, @avikivity .

File: sstables/sstable_set.cc (hunk 1 above)
Comment (tchaikov) [reply sstables/sstable_set.cc]: indeed. and dropped.

**Comment by tchaikov** [comment]
> I'm not sure we should do this: it introduces an unconditional allocation each time these methods are called and they are on a performance-sensitive path. Not super sensitive, as we are reading from the disk anyway.
> 
> I want a second opinion from the guardian of allocations, @avikivity .

ahh, this reminds me a talk by Travis on the performance of coroutine. let's wait for Avi's opinion.

**Comment by tgrabiec** [comment]
There are workloads which read from sstables often and they are in fact CPU bound.

**Comment by avikivity** [comment]
> I'm not sure we should do this: it introduces an unconditional allocation each time these methods are called and they are on a performance-sensitive path. Not super sensitive, as we are reading from the disk anyway.
> 
> I want a second opinion from the guardian of allocations, @avikivity .

Even when reading from disk, allocations aren't to be thrown around lightly (though we have much to improve) on the fast path.

Here, we might be reading from cached index pages. So we should avoid unnecessaty allocations.

**Comment by tchaikov** [comment]
thanks for your insights, gentlemen. i am closing this pull requests based on your inputs.
